### PR TITLE
mapError method and mapCatching errorTransform parameter

### PR DESCRIPTION
**Pull request by derkbell**

### CODE REVIEW

File: src/result.ts [hunk 1]
```diff
@@ -351,19 +351,91 @@ export class AsyncResult<Value, Err> extends Promise<Result<Value, Err>> {
...
 	 * Like {@linkcode AsyncResult.map} it transforms the value of a successful result using the {@link transform} callback.
 	 * In addition, it catches any exceptions that might be thrown inside the {@link transform} callback and encapsulates them
 	 * in a failed result.
 	 *
 	 * @param transform callback function to transform the value of the result. The callback can be async as well.
+	 * @param errorTransform optional callback function to transform the value of the error. The callback can be async as well.
 	 * @returns a new {@linkcode AsyncResult} instance with the transformed value
 	 */
-	mapCatching<ReturnType>(transform: (value: Value) => ReturnType) {
+	mapCatching<ReturnType, ErrorType extends AnyValue>(
```
Comment (everweij) [inline src/result.ts:429]: Atm, we're not doing anything with the `ErrorType` in the return type of `mapCatching`.
We can fallback to `NativeError` when the error transform-fn is not being used, and use the generic `ErrorType` in the return type of `mapCatching`:

```suggestion
	mapCatching<ReturnType, ErrorType extends AnyValue = NativeError>(
```

File: src/result.ts [hunk 2]
```diff
@@ -351,19 +351,91 @@ export class AsyncResult<Value, Err> extends Promise<Result<Value, Err>> {
...
+		transform: (value: Value) => ReturnType,
+		errorTransform?: (error: unknown) => ErrorType,
+	) {
 		return new AsyncResult<any, any>((resolve) => {
 			this.map(transform)
 				.then((result: AnyResult) => resolve(result))
-				.catch((error: unknown) => resolve(Result.error(error)));
+				.catch((error: unknown) =>
+					resolve(Result.error(errorTransform ? errorTransform(error) : error)),
+				);
```
Comment (everweij) [inline src/result.ts:438]: We should probably wrap the body of the `catch` in an additional try-catch, because the `errorTransform` function might throw, and atm those exceptions get swallowed. Instead, I think the caught expection should be passed to the missing `reject` callback of the `AsyncResult` constructor.

Something along the lines of:
```ts
try {
  resolve(
    Result.error(transformError ? transformError(error) : error),
  );
} catch (err) {
  reject(err);
}
```

File: src/result.ts [hunk 3]
```diff
@@ -351,19 +351,91 @@ export class AsyncResult<Value, Err> extends Promise<Result<Value, Err>> {
...
 		}) as ReturnType extends Promise<infer PromiseValue>
 			? PromiseValue extends Result<infer ResultValue, infer ResultError>
 				? AsyncResult<ResultValue, Err | ResultError | NativeError>
```
Comment (everweij) [inline src/result.ts:441]: Suggestion to replace all ocurrences of `NativeError` in this return type with `ErrorType`

File: src/result.ts [hunk 4]
```diff
@@ -922,18 +1002,75 @@ export class Result<Value, Err> {
...
 	 * In addition, it catches any exceptions that might be thrown inside the {@link transform} callback and encapsulates them
 	 * in a failed result.
 	 *
 	 * @param transform callback function to transform the value of the result. The callback can be async as well.
+	 * @param errorTransform optional callback function to transform the value of the error. The callback can be async as well.
 	 * @returns a new {@linkcode Result} instance with the transformed value, or a new {@linkcode AsyncResult} instance
 	 * if the transform function is async.
 	 */
-	mapCatching<ReturnType>(transform: (value: Value) => ReturnType) {
+	mapCatching<ReturnType, ErrorType extends AnyValue>(
```
Comment (everweij) [inline src/result.ts:1063]: I think the same comments about the async-variant apply here as well

File: src/result.ts [hunk 5]
```diff
@@ -922,18 +1002,75 @@ export class Result<Value, Err> {
...
+		transform: (value: Value) => ReturnType,
+		errorTransform?: (error: unknown) => ErrorType,
+	) {
 		return (
-			this.success ? Result.try(() => transform(this._value)) : this
+			this.success
+				? Result.try(
+						() => transform(this._value),
+						(error: any) => (errorTransform ? errorTransform(error) : error),
```
Comment (everweij) [inline src/result.ts:1071]: If I'm correct, you can pass the reference to `errorTransform` fn as is, because they both have the same signature.

File: src/result.ts [hunk 6]
```diff
@@ -351,19 +351,91 @@ export class AsyncResult<Value, Err> extends Promise<Result<Value, Err>> {
...
+				: AsyncResult<Value, Err>;
+	}
+
 	/**
 	 * Like {@linkcode AsyncResult.map} it transforms the value of a successful result using the {@link transform} callback.
 	 * In addition, it catches any exceptions that might be thrown inside the {@link transform} callback and encapsulates them
 	 * in a failed result.
 	 *
 	 * @param transform callback function to transform the value of the result. The callback can be async as well.
+	 * @param errorTransform optional callback function to transform the value of the error. The callback can be async as well.
```
Comment (everweij) [inline src/result.ts:426]: I don't think we should support an async `errorTransform` (at least for now), and I don't think this implementation supports it either.

File: src/result.test.ts [hunk 7]
```diff
@@ -1198,6 +1198,147 @@ describe("Result", () => {
 			});
 		});
 
+		describe("mapError", () => {
```
Comment (everweij) [inline src/result.test.ts:1201]: I see that you use `CustomError` a lot in the added test cases. At first glance there is nothing particularly wrong with this, although type-wise it tend to 'slip through the cracks' here and there. Are you ok with using the `ErrorA` and `ErrorB` classes to proof things like transformations of errors, etc? As bonus you only have to match against the constructor of the error.

File: src/result.test.ts [hunk 8]
```diff
@@ -1235,6 +1376,22 @@ describe("Result", () => {
...
+
+			it("transforms thrown errors when an errorTransform function is defined", async () => {
+				const result: Result<number, CustomError> = await Result.ok(
+					2,
+				).mapCatching(
+					async (): Promise<number> => {
+						throw new Error("TEST_ERROR");
+					},
+					(_error) => {
+						return new CustomError();
```
Comment (everweij) [inline src/result.test.ts:1388]: This is an example of where `CustomError` creates a false positive.
If you replace this with `ErrorA` it will probably fail the typecheck. See comment about types of `mapCatching` in `result.ts`.

File: src/result.test.ts [hunk 9]
```diff
@@ -1235,6 +1376,22 @@ describe("Result", () => {
...
+					},
+				);
+
+				Result.assertError(result);
+				expect(result.error).toBeInstanceOf(CustomError);
+			});
 		});
```
Comment (everweij) [inline src/result.test.ts:1395]: Can you add a test for when an exception gets thrown inside the `errorTransform` fn?

File: src/result.test.ts [hunk 10]
```diff
@@ -1980,6 +2250,20 @@ describe("AsyncResult", () => {
 				Result.assertError(result);
 				expect(result.error).toBeInstanceOf(CustomError);
 			});
+
+			it("transforms thrown errors when an errorTransform function is defined", async () => {
```
Comment (everweij) [inline src/result.test.ts:2254]: I think the same comments of the sync-variant of `mapCatching` apply here as well

**Review by everweij — CHANGES_REQUESTED** [review]
Thanks for your work, I appreciate it!! 👍🏻
I think in general it looks good. I have a couple of comments, and one bigger concern in general: are there any practical use cases where it makes sense for `mapError` to accept anything other than a function that returns only another error? I notice that in this PR `mapError` also accepts the transform fn to be async and/or return another `Result`, and I'm not sure about that. I could be overlooking something, so please correct me, but in general I like to restrict the API as much as possible, with the reasoning that it is always easier to relax it later on, instead of vice versa.
Let me know what your thoughts are, or if you need any help.

**Comment by derkbell** [comment]
Yeah, it could be simplified. I was just going to describe a particular use case I have/had. The fetch API returns a Response object that can be checked for an ok-ish response (2xx status code) by checking the response.ok boolean. In case of an ok response, I expect response.data to be valid JSON. In case of a 500 error, it could be valid JSON but it could be some raw error message from whatever failed. So in the mapError I have to do a response.json() which is async and could throw as well.

However, I think that could also be handled with a map because map can return a Result. Something like
```typescript
return await Result.try(
    () => fetch(url, options),
    (error) => MyApplicationError.from(error, url),
).map(async (response) =>
    response.ok ? Result.ok(await response.json()) : Result.error(await handleErrorResponse(response)),
)
```

handleErrorResponse does await response.json() and if that fails it does await response.text(). It is (response: Response) => Promise<MyApplicationError>

So, long story short, I do have errors that need some async work but it isn't really error mapping, more error construction that can be done in a regular map

**Comment by derkbell** [comment]
On the other notes, thanks for the feedback, this is a nice way to dive into the typescript typesystem. Will go through the feedback whenever I have time this week.

**Comment by everweij** [comment]
I think it's an interesting case you got there. I will give it some further thought, but for now, I agree with your conclusion -> differentiate between error mapping and construction.

I already had a branch locally a couple of days ago that contained some drafts regarding mapError and I couldn't resist working on it some more tonight, so essentially, I just pushed what I've got and added the things that I discussed in the comments (https://github.com/everweij/typescript-result/pull/9).
I know you have put quite some effort into this PR, and I almost don't dare to ask, but are you ok with picking my PR over yours?

**Comment by derkbell** [comment]
Yeah go for it, don't worry, it was a good practice session for me. Learned a lot (and switching to biome for formatting right now)

**Comment by everweij** [comment]
Ok, that's good to hear! Thanks again for your help and your input. Don't hesitate to submit another issue if you need help or have any suggestions to improve this lib.
